[CK_TILE] Pooling FWD (Lwpck 3683)

amd-yashagar opened the pull request:
## Proposed changes

This PR implements the forward 2D/3D pooling kernel in ck_tile. It uses the existing 2D reduce kernel primitives for the implementation.  3D pooling example and necessary tests have also been added.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x ] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ x] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [ x] I have run `clang-format` on all changed files
- [ ] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

amd-yashagar committed:
Pooling 2D/3D with refernce

amd-yashagar committed:
Tests & cleanup

- added test for ppoling
- cleanup
- removed 2d example

AviralGoelAMD commented on example/ck_tile/36_pooling/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+set(EXAMPLE_POOL_3D "example_pool3d")
```
we follow the convention ck_tile_example_EXAMPLE_NAME

AviralGoelAMD commented on example/ck_tile/36_pooling/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+target_compile_options(${EXAMPLE_POOL_3D} PRIVATE ${EXAMPLE_POOL_COMPILE_OPTIONS})
+
+set_property(GLOBAL PROPERTY RULE_MESSAGES OFF)
```
```suggestion
```

I don't think we need it. For examples like 03_gemm we did not use it.

AviralGoelAMD commented on example/ck_tile/36_pooling/pool3d.cpp:
```diff
@@ -0,0 +1,191 @@
+
+#include "ck_tile/host.hpp"
+#include "ck_tile/ops/pool.hpp"
+#include "ck_tile/host/reference/reference_pool.hpp"
```
can be included in host.hpp
```suggestion
```

AviralGoelAMD commented on example/ck_tile/36_pooling/pool3d.cpp:
```diff
@@ -0,0 +1,191 @@
+#include <cstring>
+
+// Parse command-line arguments for 3D pooling example
+auto create_args(int argc, char* argv[])
```
Please add README.md to describe usage

AviralGoelAMD commented on example/ck_tile/36_pooling/pool3d.cpp:
```diff
@@ -0,0 +1,191 @@
+auto create_args(int argc, char* argv[])
+{
+    ck_tile::ArgParser arg_parser;
+    arg_parser.insert("N", "2", "N dimension")
```
mention default values in description

AviralGoelAMD commented on example/ck_tile/36_pooling/pool3d.cpp:
```diff
@@ -0,0 +1,191 @@
+    // Validate kernel can handle the given configuration
+    if(!Kernel::IsSupportedArgument(kernel_args))
+    {
+        std::cout << "ERROR: Kernel arguments are not supported!" << std::endl;
```
throw std::runtime_error("Wrong! Arguments not supported!\n");

AviralGoelAMD commented on test/ck_tile/pooling/test_pooling.cpp:
```diff
@@ -0,0 +1,213 @@
+TYPED_TEST(TestCkTilePooling, Pool3D_2x2x2)
+{
+    typename TestFixture::Config3D config = {
+        1, 4, 4, 4, 32, 2, 2, 2, 2, 2, 2, 1, 1, 1, 0, 0, 0, 0, 0, 0, "2x2x2 pooling"};
```
Can you use comments and clang format off to annotate each value?

illsilin committed:
Merge branch 'develop' into LWPCK-3683

amd-yashagar committed:
Comment resolution

- README added
- example target name rectified
- appropriate arg description and comments added

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3683

amd-yashagar committed:
clang-format

amd-yashagar committed:
appropriate blocksize calc

amd-yashagar committed:
modifications for future indexing addition

- instead of transforming views we now transform the descriptors, so
that the same descriptor can be re-used for index tensor in the future

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3683

amd-yashagar committed:
some basic fixes

aosewski commented on example/ck_tile/36_pooling/pool3d.cpp:
```diff
@@ -0,0 +1,190 @@
+    // M dimension = merged (N,Do,Ho,Wo,C)
+    const ck_tile::index_t M = N * Do * Ho * Wo * C;
+    const ck_tile::index_t kGridSize =
+        (M + BlockTile::at(ck_tile::number<0>{}) - 1) / BlockTile::at(ck_tile::number<0>{});
```
The Kernel should calculate this.

aosewski commented on include/ck_tile/ops/pooling/pipeline/pool_shape.hpp:
```diff
@@ -0,0 +1,44 @@
+    static constexpr index_t WarpPerBlock_N = BlockWarps::at(number<1>{});
+
+    static constexpr index_t RepeatInWarp =
+        Warp_M * Warp_N / ThreadTile_M / ThreadTile_N / ck_tile::get_warp_size();
```
You should assert that `Warp_M * Warp_N / ThreadTile_M / ThreadTile_N` is a multiple of warp size.

amd-yashagar committed:
comment resolutions

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3683

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3683

AviralGoelAMD commented on include/ck_tile/ops/pooling/kernel/pool_kernel.hpp:
```diff
@@ -0,0 +1,496 @@
+};
+
+template <typename Problem_, typename Policy_ = PoolDefaultPolicy>
+struct Pool
```
the convention we follow is struct OperationKernel
```suggestion
struct PoolKernel
```

AviralGoelAMD commented on include/ck_tile/ops/pooling/kernel/pool_kernel.hpp:
```diff
@@ -0,0 +1,496 @@
+                                        make_tuple(sequence<0, 2, 4, 5>{}, sequence<1, 3>{}),
+                                        make_tuple(sequence<0>{}, sequence<1>{}));
+
+        const auto in_desc_padded = transform_tensor_descriptor(
```
It would cool idea to visually depict the transformations in the README.md (optionally)

AviralGoelAMD commented on include/ck_tile/ops/pooling/kernel/pool_kernel.hpp:
```diff
@@ -0,0 +1,496 @@
+    }
+
+    template <typename TensorShape, typename WindowShape>
+    static CK_TILE_DEVICE auto MakeView2D(PoolKernelArgs<TensorShape, WindowShape> kargs)
```
```suggestion
    static CK_TILE_DEVICE auto MakeTensorView2D(PoolKernelArgs<TensorShape, WindowShape> kargs)
```

I'd recommend calling it this just for code readability.

amd-yashagar committed:
comment resolutions

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3683
